### PR TITLE
A little more cleanup and updating models

### PR DESCRIPTION
**Pull request by joaoburatto**
Update on the walls models, using the lighter color one.
Update on the human player model.
Updated prefabs, locker, flashlight, crowbar.
Temporary fix to the floating problem while walking.
Changed the "Body" script so the severed limbs have the same material as the ones on the origin.
Added the limb cap texture, and updated the color on the player material.
Added a simple collision layer, Items, that just avoid items colliding with each other.

### CODE REVIEW

File: Assets/Scripts/Player/Body/Body.cs [hunk 1]
```diff
@@ -101,5 +106,36 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
+            newBodyPartMesh.materials = materials;
+
+            // there's 5 shape keys on the human model
+            for (int i = 0; i < 4; i++)
+            {
+                newBodyPartMesh.SetBlendShapeWeight(i, bodyPartMesh.GetBlendShapeWeight(i));
+            }
+        }
+
+        private void Start()
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Body/Body.cs]: Move Start() to the top.

File: Assets/Scripts/Player/Body/Body.cs [hunk 2]
```diff
@@ -101,5 +106,36 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
 		}
-	}
+
+        private void UpdateBodyPartList()
+        {
+            bodyParts.Clear();
+            foreach (BodyPart part in gameObject.GetComponentsInChildren<BodyPart>())
+            {
+                bodyParts.Add(part);
+
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Body/Body.cs]: unnecessary whitespace

File: Assets/Scripts/Player/Body/Body.cs [hunk 3]
```diff
@@ -101,5 +106,36 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
+        {
+            // Making sure the part is equal to the original one
+            SkinnedMeshRenderer newBodyPartMesh = newBodyPart.GetComponent<SkinnedMeshRenderer>();
+            SkinnedMeshRenderer bodyPartMesh = bodyPart.SkinnedMeshRenderer;
+
+            Material[] materials = bodyPartMesh.sharedMaterials;
+            newBodyPartMesh.materials = materials;
+
+            // there's 5 shape keys on the human model
+            for (int i = 0; i < 4; i++)
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Body/Body.cs]: Any way to get the number dynamically? It would be a shame if merely updating the model required code changes.

File: Assets/Scripts/Player/Body/BodyPart.cs [hunk 4]
```diff
@@ -35,10 +35,19 @@ public class BodyPart : MonoBehaviour
...
+            }
+            if (body == null)
+            {
+                body = transform.root.GetComponent<Body>();
+            }
             if (severedBodyPartPrefab != null && severedBodyPartPrefab.GetComponent<SeveredBodyPart>() == null)
             {
                 Debug.LogError($"SeveredBodyPartPrefab on BodyPart {gameObject.name} on Body {body.gameObject.name} is missing a SeveredBodyPart component!");
             }
+
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Player/Body/BodyPart.cs]: remove whitespace

File: Assets/Scripts/Utilities/MaterialChanger.cs [hunk 5]
```diff
@@ -7,6 +7,12 @@ public class MaterialChanger : MonoBehaviour
     public Material[] cachedMaterial;
     public MeshRenderer cachedRenderer;
 
+    public  Material Palette01            = Resources.Load<Material>("Materials/Pallete01.mat");
+    public  Material Palette05            = Resources.Load<Material>("Materials/Pallete05.mat");
+    public  Material Palette05Emission    = Resources.Load<Material>("Materials/Pallete05 Emission.mat");
+    
+
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Utilities/MaterialChanger.cs]: too much whitespace

File: Assets/Scripts/Player/Body/Body.cs (hunk 3 above)
Comment (joaoburatto) [reply Assets/Scripts/Player/Body/Body.cs]: apparently, no.

File: Assets/Scripts/Player/Body/Body.cs [hunk 6]
```diff
@@ -101,5 +110,30 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: Just don't hardcode the number of blendshapes

File: Assets/Scripts/Player/Body/Body.cs [hunk 7]
```diff
@@ -101,5 +110,30 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
 		{
 			return bodyParts.FirstOrDefault(bodyParts => bodyParts.BodyPartType == bodyPartType);
 		}
-	}
+
+        private void UpdateBodyPartList()
+        {
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: This is only called once. Just have it in start and make a clear comment

File: Assets/Scripts/Player/Body/Body.cs [hunk 8]
```diff
@@ -101,5 +110,30 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
+            bodyParts.Clear();
+            foreach (BodyPart part in gameObject.GetComponentsInChildren<BodyPart>())
+            {
+                bodyParts.Add(part);
+            }
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: `bodyParts.AddRange(gameObject.GetComponentsInChildren<BodyPart>());`

File: Assets/Scripts/Player/Body/Body.cs [hunk 9]
```diff
@@ -101,5 +110,30 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
+            foreach (BodyPart part in gameObject.GetComponentsInChildren<BodyPart>())
+            {
+                bodyParts.Add(part);
+            }  
+        }
+
+        private void UpdateBodyPartVisuals(GameObject newBodyPart, BodyPart bodyPart)
+        {
+            // Making sure the part is equal to the original one
+            SkinnedMeshRenderer newBodyPartMesh = newBodyPart.GetComponent<SkinnedMeshRenderer>();
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: Just parsing in `SkinnedMeshRenderer newBodyPartMeshRenderer` is cleaner than parsing in a GameObject and getting the component. Or maybe you actually want to have it `BodyPart` like the other parameter is and the name suggests.
With you current style you have no safety and you aren't checking if the component exist.

File: Assets/Scripts/Player/Body/Body.cs [hunk 10]
```diff
@@ -101,5 +110,30 @@ private BodyPart FindBodyPart(BodyPartType bodyPartType)
...
+            SkinnedMeshRenderer bodyPartMesh = bodyPart.SkinnedMeshRenderer;
+
+            Material[] materials = bodyPartMesh.sharedMaterials;
+            newBodyPartMesh.materials = materials;
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: This feels dirty to me. This looks like initialization that should happen in `Awake()` or `Start()`. Remember that materials need to be disposed manually after being allocated, for example in `OnDestroy()`

File: Assets/Scripts/Player/Body/Body.cs [hunk 11]
```diff
@@ -14,10 +14,14 @@ public class Body : NetworkBehaviour
 	{
 		///The BodyPart objects that this MonoBehaviour is responsible for managing
 		[SerializeField] private List<BodyPart> bodyParts = new List<BodyPart>();
```
Comment (Dyhr) [inline Assets/Scripts/Player/Body/Body.cs]: If you aren't using this in the inspector then we might as well not have `[SerializeField]` on there

File: Assets/Scripts/Utilities/MaterialChanger.cs [hunk 12]
```diff
@@ -7,6 +7,11 @@ public class MaterialChanger : MonoBehaviour
     public Material[] cachedMaterial;
     public MeshRenderer cachedRenderer;
 
+    //todo: make sure the main materials are here, Unity's material manipulation is not really nice.
+    public  Material Palette01            = Resources.Load<Material>("Materials/Pallete01.mat");
+    public  Material Palette05            = Resources.Load<Material>("Materials/Pallete05.mat");
+    public  Material Palette05Emission    = Resources.Load<Material>("Materials/Pallete05 Emission.mat");
```
Comment (Dyhr) [inline Assets/Scripts/Utilities/MaterialChanger.cs]: You added these, but are not using them

File: Assets/Scripts/Player/MovementController.cs [hunk 13]
```diff
@@ -188,7 +188,31 @@ void Update()
                     currentMovement.y * Vector3.Cross(camera.transform.right, Vector3.up).normalized +
                     currentMovement.x * Vector3.Cross(Vector3.up, camera.transform.forward).normalized;
                 // Move (without gravity). Whenever we move we also readjust the player's direction to the direction they are running in.
-                characterController.Move(absoluteMovement * Time.deltaTime);
+
+                //temporary to keep us from being stuck in the air
+                Tile _currentTile = null;
```
Comment (Dyhr) [inline Assets/Scripts/Player/MovementController.cs]: Name should be `currentTile`

File: Assets/Scripts/Player/MovementController.cs [hunk 14]
```diff
@@ -188,7 +188,31 @@ void Update()
...
+
+                try
+                {
+                    _currentTile = GameObject.Find(string.Format("tile_{0}_{1}", Mathf.Round(transform.position.x), Mathf.Round(transform.position.z))).GetComponent<Tile>();
+                }
+                catch
+                {
+
+                }
```
Comment (Dyhr) [inline Assets/Scripts/Player/MovementController.cs]: I have no idea what you are trying to catch here, so please don't

File: Assets/Scripts/Player/MovementController.cs [hunk 15]
```diff
@@ -188,7 +188,31 @@ void Update()
...
                     currentMovement.x * Vector3.Cross(Vector3.up, camera.transform.forward).normalized;
                 // Move (without gravity). Whenever we move we also readjust the player's direction to the direction they are running in.
-                characterController.Move(absoluteMovement * Time.deltaTime);
+
+                //temporary to keep us from being stuck in the air
+                Tile _currentTile = null;
+
+                try
+                {
+                    _currentTile = GameObject.Find(string.Format("tile_{0}_{1}", Mathf.Round(transform.position.x), Mathf.Round(transform.position.z))).GetComponent<Tile>();
```
Comment (Dyhr) [inline Assets/Scripts/Player/MovementController.cs]: Do not use GameObject.Find, especially not in `Update()`. It is a slow method to call, and hard coding the name is very inflexible way of coding.

Comment (Dyhr) [reply Assets/Scripts/Player/MovementController.cs]: Also you can use string interpolation instead of `string.Format` for better readability

File: Assets/Scripts/Player/MovementController.cs [hunk 16]
```diff
@@ -188,7 +188,31 @@ void Update()
...
+                }
+                catch
+                {
+
+                }
+                Vector3 _gravity;
```
Comment (Dyhr) [inline Assets/Scripts/Player/MovementController.cs]: Should be named `gravity`

File: Assets/Scripts/Player/MovementController.cs [hunk 17]
```diff
@@ -188,7 +188,31 @@ void Update()
...
+                if (_currentTile == null || _currentTile.turf == null)
+                {
+                    _gravity = new Vector3(0, 0, 0);
+                    characterAnimator.SetBool("Floating", true);
+                }
+                else
+                {
+                    _gravity = new Vector3(0, 9.8f, 0);
+                    characterAnimator.SetBool("Floating", false);
+                }
```
Comment (Dyhr) [inline Assets/Scripts/Player/MovementController.cs]: This section can be shortened to something like:
```
bool floating = _currentTile == null || _currentTile.turf == null;
Vector3 _gravity = floating ? Vector3.zero : Vector3.down * 9.8f;
                
characterAnimator.SetBool("Floating", true);
```
To keep the code short for doing something this simple. For easy access you may also want to define gravity as a constant.

File: Assets/Scripts/Player/Body/Body.cs (hunk 6 above)
Comment (Dyhr) [reply Assets/Scripts/Player/Body/Body.cs]: Just saw the comment above. The answer is `for (int i = 0; i < newBodyPartMesh.sharedMesh.blendShapeCount; i++)`

**Comment by Em3rgencyLT** [comment]
Dayum @Dyhr! Amazing review as always! I feel silly now for missing all of these...

File: Assets/Scripts/Player/Body/Body.cs (hunk 6 above)
Comment (joaoburatto) [reply Assets/Scripts/Player/Body/Body.cs]: wasn't aware of that. thanks.

File: Assets/Scripts/Player/Body/Body.cs (hunk 11 above)
Comment (joaoburatto) [reply Assets/Scripts/Player/Body/Body.cs]: It was being used, I'll change now as I automated it

File: Assets/Scripts/Utilities/MaterialChanger.cs (hunk 12 above)
Comment (joaoburatto) [reply Assets/Scripts/Utilities/MaterialChanger.cs]: yes didn't wanted to change the material folder to resources right now, when we implement more item usage it will be more important
